expression/builtin_compare.go

```diff
@@ -1023,81 +1024,135 @@ func isTemporalColumn(expr Expression) bool {
 }
 
 // tryToConvertConstantInt tries to convert a constant with other type to a int constant.
-func tryToConvertConstantInt(ctx sessionctx.Context, con *Constant) *Constant {
+func tryToConvertConstantInt(ctx sessionctx.Context, isUnsigned bool, con *Constant) (_ *Constant, isOverflow bool) {
 	if con.GetType().EvalType() == types.ETInt {
-		return con
+		return con, false
 	}
 	dt, err := con.Eval(nil)
 	if err != nil {
-		return con
+		return con, false
 	}
 	sc := ctx.GetSessionVars().StmtCtx
-	i64, err := dt.ToInt64(sc)
+	fieldType := types.NewFieldType(mysql.TypeLonglong)
+	if isUnsigned {
+		fieldType.Flag |= mysql.UnsignedFlag
+	}
+	dt, err = dt.ConvertTo(sc, fieldType)
 	if err != nil {
-		return con
+		return con, terror.ErrorEqual(err, types.ErrOverflow)
 	}
 	return &Constant{
-		Value:        types.NewIntDatum(i64),
-		RetType:      types.NewFieldType(mysql.TypeLonglong),
+		Value:        dt,
+		RetType:      fieldType,
 		DeferredExpr: con.DeferredExpr,
-	}
+	}, false
 }
 
-// RefineConstantArg changes the constant argument to it's ceiling or flooring result by the given op.
-func RefineConstantArg(ctx sessionctx.Context, con *Constant, op opcode.Op) *Constant {
+// RefineComparedConstant changes the constant argument, type of which is not int, to it's ceiling or flooring result by the given op.
+// isAlwaysFalse indicates whether int column <cmp> `con` is definite false.
```

Comment:
isAlwaysFalse indicates whether the int column <cmp> "con" is false.

Reply:
For `con`, if you use the inline code format, it will not be displayed in the code comments as expected. Therefore, I use "con".